File: src/tmux_context.rs
use std::io::Error;
use std::process::Output;

use crate::tmux;

pub struct TmuxContext {
    detached_session: String,
    session: String,
    window: usize,
    pane: usize
}

pub fn create_tmux_context(detached_session: String) -> Result<TmuxContext, Error> {
    let session = match String::from_utf8(tmux::current_session()?.stdout) {
        Ok(val) => val.replace("\n", ""),
        Err(e) => panic!("Error: Could not retrieve tmux session id: {}", e)
    };
    let window = match String::from_utf8(tmux::current_window()?.stdout) {
        Ok(val) => val.replace("\n", ""),
        Err(e) => panic!("Error: Could not retrieve tmux window id: {}", e)
    };
    let pane = match String::from_utf8(tmux::current_pane()?.stdout) {
        Ok(val) => val.replace("\n", ""),
        Err(e) => panic!("Error: Could not retrieve tmux pane id: {}", e)
    };

    let window_id = match window.parse() {
        Ok(i) => i,
        Err(e) => panic!("Error: Failed to parse tmux window {}: {}", window, e)
    };
    let pane_id = match pane.parse() {
        Ok(i) => i,
        Err(e) => panic!("Error: Failed to parse tmux pane {}: {}", pane, e)
    };

    Ok(TmuxContext {
        detached_session,
        session,
        window: window_id,
        pane: pane_id,
    })
}

impl TmuxContext {
    pub fn prepare(&self) -> Result<Output, Error> {
        tmux::start_detached_session(&self.detached_session)?;
        tmux::set_remain_on_exit(&self.session, self.window, true)
    }

    pub fn cleanup(&self) -> Result<Output, Error> {
        tmux::kill_session(&self.detached_session)?;
        tmux::set_remain_on_exit(&self.session, self.window, false)
    }

    pub fn break_pane(&self, source_pane: usize, dest_window: usize, window_label: &str) -> Result<Output, Error> {
        tmux::break_pane(
            &self.session,
            self.window,
            source_pane,
            &self.detached_session,
            dest_window,
            window_label)?;
        tmux::set_remain_on_exit(&self.detached_session, dest_window, true)
    }

    pub fn join_pane(&self, target_window: usize) -> Result<usize, Error> {
        tmux::join_pane(
            &self.detached_session,
            target_window,
            &self.session,
            self.window,
            self.pane
        )?;
        Ok(self.pane + 1)
    }

    pub fn create_pane(&self, command: &str) -> Result<usize, Error> {
        let pane = tmux::create_pane(&self.session, self.window, self.pane, command)?;

        match String::from_utf8(pane.stdout) {
            Ok(val) => match val.replace("\n", "").parse() {
                Ok(i) => Ok(i),
                Err(_) => Err(Error::new(
                    std::io::ErrorKind::Other,
                    "Error: Could not convert create_pane output to int"
                ))
            },
            Err(_) => Err(Error::new(std::io::ErrorKind::Other, "Error: Could not parse create_pane output"))
        }
    }
}

File: internal/proctmux/tmux_context.go
package proctmux

import (
	"fmt"
	"log"
	"os"
	"os/exec"
	"slices"
	"strconv"
	"strings"
)

type TmuxContext struct {
	PaneID            string
	SessionID         string
	DetachedSessionID string
	SplitSize         string
}

func NewTmuxContext(detachedSession string, killExistingSession bool, processListWidthPercent int) (*TmuxContext, error) {
	paneID, err := CurrentPane()
	if err != nil {
		return nil, fmt.Errorf("could not retrieve tmux pane id: %w", err)
	}
	sessionID, err := CurrentSession()
	if err != nil {
		return nil, fmt.Errorf("could not retrieve tmux session id: %w", err)
	}
	sessions, err := ListSessions()
	if err != nil {
		return nil, fmt.Errorf("could not list tmux sessions: %w", err)
	}
	detachedSessionID := ""

	exists := slices.Contains(sessions, detachedSession)
	if exists {
		log.Printf("Session '%s' already exists", detachedSession)
		if killExistingSession {
			if err := KillSession(detachedSession); err != nil {
				return nil, fmt.Errorf("could not kill existing session: %w", err)
			}
			id, err := StartDetachedSession(detachedSession)
			if err != nil {
				return nil, fmt.Errorf("could not start detached session after kill: %w", err)
			}
			detachedSessionID = id
		} else {
			return nil, fmt.Errorf("session '%s' already exists (set killExistingSession to true to replace it)", detachedSession)
		}
	} else {
		id, err := StartDetachedSession(detachedSession)
		if err != nil {
			return nil, fmt.Errorf("could not start detached session: %w", err)
		}
		detachedSessionID = id
	}
	splitSize := 100 - processListWidthPercent

	return &TmuxContext{
		PaneID:            paneID,
		SessionID:         sessionID,
		DetachedSessionID: detachedSessionID,
		SplitSize:         strconv.Itoa(splitSize) + "%",
	}, nil
}

// buildEnvWithAddPath merges process Env and AddPath into a final env map.
// It appends AddPath entries to PATH (matching Python's sys.path.append semantics).
func buildEnvWithAddPath(p *Process) map[string]string {
	merged := map[string]string{}
	if p.Config != nil && p.Config.Env != nil {
		for k, v := range p.Config.Env {
			merged[k] = v
		}
	}
	// Base PATH from explicit env PATH or current process PATH
	path := merged["PATH"]
	if path == "" {
		path = os.Getenv("PATH")
	}
	if p.Config != nil && len(p.Config.AddPath) > 0 {
		sep := string(os.PathListSeparator)
		for _, ap := range p.Config.AddPath {
			if ap == "" {
				continue
			}
			if path == "" {
				path = ap
			} else {
				path = path + sep + ap // append
			}
		}
	}
	merged["PATH"] = path
	return merged
}

func (t *TmuxContext) CreatePane(process *Process) (string, int, error) {
	// Create a new pane split off the main proctmux pane, not the process's pane
	return CreatePane(
		t.PaneID,
		process.Command(),
		process.Config.Cwd,
		buildEnvWithAddPath(process),
		t.SplitSize,
	)
}

// Create a pane in the detached session
func (t *TmuxContext) CreateDetachedPane(process *Process) (string, int, error) {
	paneID, pid, err := CreateDetachedPane(
		t.DetachedSessionID,
		process.ID,
		process.Label,
		process.Command(),
		process.Config.Cwd,
		buildEnvWithAddPath(process),
	)
	if err != nil {
		return "", 0, err
	}
	return paneID, pid, nil
}

// Move a pane to the detached session
func (t *TmuxContext) BreakPane(paneID string, destWindow int, windowLabel string) error {
	return BreakPane(paneID, t.DetachedSessionID, destWindow, windowLabel)
}

// Move a pane from the detached session to the user session
func (t *TmuxContext) JoinPane(sourcePaneID string) error {
	return JoinPane(sourcePaneID, t.PaneID, t.SplitSize)
}

func (t *TmuxContext) Prepare() error {
	return SetGlobalRemainOnExit(true)
}

func (t *TmuxContext) Cleanup() error {
	if err := KillSession(t.DetachedSessionID); err != nil {
		return err
	}
	return SetGlobalRemainOnExit(false)
}

func (t *TmuxContext) FocusPane(paneID string) error {
	return SelectPane(paneID)
}

func (t *TmuxContext) ToggleZoom(paneID string) error {
	return ToggleZoom(paneID)
}

func (t *TmuxContext) PaneVariables(paneID, format string) (string, error) {
	args := []string{"list-panes", "-t", paneID, "-F", format}
	out, err := exec.Command("tmux", args...).Output()
	if err != nil {
		return "", err
	}
	return strings.TrimSpace(string(out)), nil
}

func (t *TmuxContext) IsZoomedIn() bool {
	out, err := t.PaneVariables(t.PaneID, "#{window_zoomed_flag} #{pane_active}")
	if err != nil {
		return false
	}
	return out == "1 1"
}

func (t *TmuxContext) ZoomIn() error {
	if !t.IsZoomedIn() {
		return t.ToggleZoom(t.PaneID)
	}
	return nil
}

func (t *TmuxContext) ZoomOut() error {
	if t.IsZoomedIn() {
		return t.ToggleZoom(t.PaneID)
	}
	return nil
}

func (t *TmuxContext) GetPanePID(paneID string) (int, error) {
	out, err := exec.Command("tmux", "display-message", "-p", "-t", paneID, "#{pane_pid}").Output()
	if err != nil {
		return 0, err
	}
	return strconv.Atoi(strings.TrimSpace(string(out)))
}

// GetPaneSessionType wraps GetPaneSessionType using the context's session IDs.
func (t *TmuxContext) GetPaneSessionType(paneID string) (string, error) {
	return GetPaneSessionType(paneID, t.SessionID, t.DetachedSessionID)
}
